src/cpp/fastdds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,722 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file DomainParticipantImpl.cpp
+ *
+ */
+
+#include "DomainParticipantImpl.hpp"
+#include <fastdds/domain/DomainParticipant.hpp>
+#include <fastdds/domain/DomainParticipantListener.hpp>
+#include <fastrtps/rtps/participant/ParticipantDiscoveryInfo.h>
+#include <fastrtps/rtps/reader/ReaderDiscoveryInfo.h>
+#include <fastrtps/rtps/writer/WriterDiscoveryInfo.h>
+#include <fastrtps/rtps/builtin/liveliness/WLP.h>
+
+#include <fastrtps/topic/TopicDataType.h>
+
+#include <fastrtps/rtps/participant/RTPSParticipant.h>
+
+#include <fastrtps/attributes/PublisherAttributes.h>
+#include "../publisher/PublisherImpl.hpp"
+#include <fastdds/publisher/Publisher.hpp>
+
+#include <fastrtps/attributes/SubscriberAttributes.h>
+#include "../subscriber/SubscriberImpl.hpp"
+#include <fastdds/subscriber/Subscriber.hpp>
+
+#include <fastdds/topic/DataReader.hpp>
+#include <fastdds/topic/DataWriter.hpp>
+
+#include <fastrtps/rtps/RTPSDomain.h>
+
+#include <fastrtps/transport/UDPv4Transport.h>
+#include <fastrtps/transport/UDPv6Transport.h>
+#include <fastrtps/transport/test_UDPv4Transport.h>
+
+#include <fastrtps/log/Log.h>
+
+#include <chrono>
+
+using namespace eprosima;
+using namespace eprosima::fastdds;
+
+using fastrtps::ParticipantAttributes;
+using fastrtps::TopicDataType;
+using fastrtps::rtps::RTPSDomain;
+using fastrtps::rtps::RTPSParticipant;
+using fastrtps::rtps::ParticipantDiscoveryInfo;
+#if HAVE_SECURITY
+using fastrtps::rtps::ParticipantAuthenticationInfo;
+#endif
+using fastrtps::rtps::ReaderDiscoveryInfo;
+using fastrtps::rtps::ReaderProxyData;
+using fastrtps::rtps::WriterDiscoveryInfo;
+using fastrtps::rtps::WriterProxyData;
+using fastrtps::rtps::InstanceHandle_t;
+using fastrtps::rtps::GUID_t;
+using fastrtps::rtps::EndpointKind_t;
+using fastrtps::rtps::ResourceEvent;
+using fastrtps::Log;
+
+DomainParticipantImpl::DomainParticipantImpl(
+        const ParticipantAttributes& patt,
+        DomainParticipant* pspart,
+        DomainParticipantListener* listen)
+    : att_(patt)
+    , rtps_participant_(nullptr)
+    , participant_(pspart)
+    , listener_(listen)
+#pragma warning (disable : 4355 )
+    , rtps_listener_(this)
+{
+    participant_->impl_ = this;
+}
+
+void DomainParticipantImpl::disable()
+{
+    participant_->set_listener(nullptr);
```

Comment:
The standard doesn't say anything about `disable` method neither in DomainEntity or DomainParticipant if I'm not wrong. That method isn't public (it is only defined in the Impl file) for the internal cleanup process prior to destruction, so nobody and never we will be enabled after `disable` is called, since we will be destroyed.

Reply:
Understood. But what about `DataReader::disable()` and `DataWriter::disable()` (not sure if there are others)? Those are public to the user, aren't they?

Reply:
`DataReader::disable()` and `DataWriter::disable()` doesn't exists now.